cql3/functions/functions.hh

```diff
@@ -89,19 +87,19 @@ private:
                               shared_ptr<function> fun,
                               const std::vector<shared_ptr<assignment_testable>>& provided_args,
                               const sstring& receiver_ks,
-                              std::optional<const std::string_view> receiver_cf);
+                              std::optional<const std::string_view> receiver_cf) const;
     assignment_testable::test_result match_arguments(data_dictionary::database db, const sstring& keyspace,
             const schema* schema_opt,
             shared_ptr<function> fun,
             const std::vector<shared_ptr<assignment_testable>>& provided_args,
             const sstring& receiver_ks,
-            std::optional<const std::string_view> receiver_cf);
+            std::optional<const std::string_view> receiver_cf) const;
 
-    bool type_equals(const std::vector<data_type>& t1, const std::vector<data_type>& t2);
+    bool type_equals(const std::vector<data_type>& t1, const std::vector<data_type>& t2) const;
 };
 
 // Getter for static functions object.
-functions& instance();
+const functions& instance();
 
 class change_batch : public functions {
```

Comment:
I'm still not sure whether `change_batch` will be needed anymore. To me, it looks like you could just use non-const `function` instances, i.e. modify them and ultimately set the thread-local instance via a setter function.

However, I don't have a clear picture of how your code will ultimately look like and whether it would be possible to achieve what you want to achieve with the pattern I propose. Therefore, I just leave this as a suggestion here, this comment is not a blocker.

Reply:
I want to avoid here storing copy in a static variable, because then you need to care about concurrent usages and also about cleaning (e.g. after exception). That's why this class was added.

Reply:
Maybe making functions a member of change_batch would be better, but it's marginal at this point.